queue/config.go

```diff
@@ -25,18 +29,118 @@ func DefaultConfig() Config {
+	UseEncryption      bool              `config:"use_encryption"`
+	UseCompression     bool              `config:"use_compression"`
+}
+
+func (c *DiskConfig) Validate() error {
+	// If the segment size is explicitly specified, the total queue size must
+	// be at least twice as large.
+	if c.SegmentSize != nil && c.MaxSize != 0 && c.MaxSize < *c.SegmentSize*2 {
+		return fmt.Errorf("disk queue max_size must be at least twice as big as segment_size")
+	}
+
+	// We require a total queue size of at least 10MB, and a segment size of
+	// at least 1MB. The queue can support lower thresholds, but it will perform
+	// terribly, so we give an explicit error in that case.
+	// These bounds are still extremely low for Beats ingestion, but if all you
+	// need is for a low-volume stream on a tiny device to persist between
+	// restarts, it will work fine.
+	if c.MaxSize != 0 && c.MaxSize < 10*1000*1000 {
+		return fmt.Errorf(
+			"disk queue max_size (%d) cannot be less than 10MB", c.MaxSize)
+	}
+	if c.SegmentSize != nil && *c.SegmentSize < 1000*1000 {
+		return fmt.Errorf(
+			"disk queue segment_size (%d) cannot be less than 1MB", *c.SegmentSize)
+	}
+
+	if c.RetryInterval != nil && c.MaxRetryInterval != nil &&
+		*c.MaxRetryInterval < *c.RetryInterval {
+		return fmt.Errorf(
+			"disk queue max_retry_interval (%v) can't be less than retry_interval (%v)",
+			*c.MaxRetryInterval, *c.RetryInterval)
+	}
+
+	if c.UseEncryption && c.EncryptionPassword == "" {
+		return fmt.Errorf("fetching encryption key is currently unsupported, set `encryption_password`")
+	}
+
+	return nil
+}
+
+func DiskSettingsFromConfig(diskConfig *DiskConfig) (diskqueue.Settings, error) {
+	settings := defaultDiskqueueSettings()
+	settings.Path = diskConfig.Path
+	settings.MaxBufferSize = uint64(diskConfig.MaxSize)
+	if diskConfig.SegmentSize != nil {
+		settings.MaxSegmentSize = uint64(*diskConfig.SegmentSize)
+	} else {
+		// If no value is specified, default segment size is total queue size
+		// divided by 10.
+		settings.MaxSegmentSize = uint64(diskConfig.MaxSize) / 10
+	}
+
+	if diskConfig.ReadAheadLimit != nil {
+		settings.ReadAheadLimit = *diskConfig.ReadAheadLimit
+	}
+
+	if diskConfig.WriteAheadLimit != nil {
+		settings.WriteAheadLimit = *diskConfig.WriteAheadLimit
+	}
+
+	if diskConfig.RetryInterval != nil {
+		settings.RetryInterval = *diskConfig.RetryInterval
+	}
+	if diskConfig.MaxRetryInterval != nil {
+		settings.MaxRetryInterval = *diskConfig.RetryInterval
+	}
+
+	if diskConfig.UseEncryption && diskConfig.EncryptionPassword != "" {
+		// Use pbkdf2 to convert the string to a key of correct size
+		settings.EncryptionKey = pbkdf2.Key([]byte(diskConfig.EncryptionPassword),
+			[]byte(diskConfig.EncryptionPassword),
+			4096,
+			diskqueue.KeySize,
+			sha1.New)
+	}
+
+	if diskConfig.UseEncryption && diskConfig.EncryptionPassword == "" { //nolint:staticcheck // TODO
+		// TODO fetch key from elastic-agent
```

Comment:
Can you create an issue for this if there isn't one, and link it from here?

Reply:
Although, rather than requesting the config it may be simpler to have the agent just populate the `EncryptionPassword` key for us on startup.

The shipper doesn't/shouldn't write the configuration received from the agent to disk, it just lives in memory (it is written to disk when collecting diagnostics, but this field should be redacted automatically since it contains the word password).

Reply:
I like `elastic-agent` just sending it.  Will go with that for now, and if we need to we can fall back to requesting the key.

Reply:
We'll need an issue on the agent side to implement this. Can you create one?

Probably we can also close https://github.com/elastic/elastic-agent-shipper/issues/121 since we won't need to change the shipper further.

Reply:
closed #121 in favor of elastic/elastic-agent#1656, updated #118 with new issue